File: src/main/java/de/pflugradts/passbird/domain/model/transfer/Input.java
package de.pflugradts.passbird.domain.model.transfer;

import de.pflugradts.passbird.domain.model.ddd.ValueObject;
import de.pflugradts.passbird.domain.model.namespace.NamespaceSlot;
import io.vavr.control.Try;

import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.FIRST;
import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.INVALID;
import static de.pflugradts.passbird.domain.model.namespace.NamespaceSlot.LAST;

/**
 * An Input represents data given by the user through the
 * {@link de.pflugradts.passbird.application.UserInterfaceAdapterPort UserInterface}.
 */
public class Input implements ValueObject {

    private final Bytes bytes;

    public static Input empty() {
        return Input.of(Bytes.emptyBytes());
    }

    private Input(final Bytes bytes) {
        this.bytes = bytes;
    }

    public static Input of(final Bytes bytes) {
        return new Input(bytes);
    }

    public Bytes getBytes() {
        return bytes;
    }

    public Bytes getCommand() {
        if (bytes.getSize() > 0) {
            for (int i = 1; i < bytes.getSize(); i++) {
                if (CharValue.Companion.charValueOf(bytes.getByte(i)).isAlphabeticCharacter()) {
                    return bytes.slice(0, i);
                }
            }
            return bytes;
        } else {
            return Bytes.emptyBytes();
        }
    }

    public Bytes getData() {
        return getBytes().getSize() > 1
                ? getBytes().slice(getCommand().getSize(), getBytes().getSize())
                : Bytes.emptyBytes();
    }

    public NamespaceSlot parseNamespace() {
        return Try.of(() -> Integer.parseInt(getBytes().asString()))
            .map(value -> value >= FIRST - 1 && value <= LAST ? NamespaceSlot.at(value) : INVALID)
            .getOrElse(INVALID);
    }

    public boolean isEmpty() {
        return bytes.isEmpty();
    }

    public void invalidate() {
        getBytes().scramble();
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/model/transfer/Input.kt
package de.pflugradts.passbird.domain.model.transfer

import de.pflugradts.passbird.domain.model.ddd.ValueObject
import de.pflugradts.passbird.domain.model.shell.PlainValue.Companion.plainValueOf
import de.pflugradts.passbird.domain.model.shell.Shell
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.emptyShell
import de.pflugradts.passbird.domain.model.slot.Slot
import de.pflugradts.passbird.domain.model.slot.Slot.Companion.FIRST_SLOT
import de.pflugradts.passbird.domain.model.slot.Slot.Companion.LAST_SLOT
import de.pflugradts.passbird.domain.model.slot.Slot.Companion.slotAt
import de.pflugradts.passbird.domain.model.slot.Slot.DEFAULT

class Input private constructor(val shell: Shell) : ValueObject {

    val command get(): Shell {
        if (shell.size > 0) {
            for (i in 1 until shell.size) {
                if (plainValueOf(shell.getByte(i)).isAlphabeticCharacter) {
                    return shell.slice(0, i)
                }
            }
            return shell
        }
        return emptyShell()
    }

    val data get() = if (shell.size > 1) shell.slice(command.size, shell.size) else emptyShell()
    val isEmpty get() = shell.isEmpty
    val isNotEmpty get() = shell.isNotEmpty
    fun invalidate() = shell.scramble()
    fun extractNestSlot(): Slot = shell.asString().toIntOrNull()?.let {
        if (it in FIRST_SLOT - 1..LAST_SLOT) slotAt(it) else DEFAULT
    } ?: DEFAULT

    override fun equals(other: Any?): Boolean = when {
        (this === other) -> true
        (javaClass != other?.javaClass) -> false
        else -> shell == (other as Input).shell
    }

    override fun hashCode() = shell.hashCode()

    companion object {
        fun inputOf(shell: Shell) = Input(shell)
        fun emptyInput() = inputOf(emptyShell())
    }
}
